sentinelhub/areas.py

```diff
@@ -475,3 +484,152 @@ def _make_split(self):
+    """
+    def __init__(self, shape_list, crs, bbox_size):
+        """
+        :param shape_list: A list of geometrical shapes describing the area of interest
+        :type shape_list: list(shapely.geometry.multipolygon.MultiPolygon or shapely.geometry.polygon.Polygon)
+        :param crs: Coordinate reference system of the shapes in `shape_list`
+        :type crs: CRS
+        :param bbox_size: Physical size in metres of generated bounding boxes. Could be a float or tuple of floats
+        :type bbox_size: int or (int, int) or float or (float, float)
+        """
+        super().__init__(shape_list, crs)
+
+        self.bbox_size = self._parse_split_parameters(bbox_size, allow_float=True)
+
+        self.shape_geometry = Geometry(self.area_shape, self.crs).transform(CRS.WGS84)
+
+        self.utm_grid = self._get_utm_polys()
+
+        self._make_split()
+
+    @abstractmethod
+    def _get_utm_polys(self):
+        raise NotImplementedError
+
+    @staticmethod
+    def _get_utm_from_props(utm_dict):
+        """ Return the UTM CRS corresponding to the UTM described by the properties dictionary
+
+        :param utm_dict: Dictionary reporting name of the UTM zone and MGRS grid
+        :type utm_dict: dict
+        :return: UTM coordinate reference system
+        :rtype: sentinelhub.CRS
+        """
+        return CRS('32{}{}'.format(6 if utm_dict['direction'] == 'N' else 7, str(utm_dict['zone']).zfill(2)))
+
+    def _align_bbox_to_size(self, bbox):
+        """ Align input bbox coordinates to be multiples of the bbox size
+
+        :param bbox: Bounding box in UTM coordinates
+        :type bbox: sentinelhub.BBox
+        :return: BBox objects with coordinates multiples of the bbox size
+        :rtype: sentinelhub.BBox
+        """
+        size_x, size_y = self.bbox_size
+        lower_left_x, lower_left_y = bbox.get_lower_left()
+        return BBox([(lower_left_x - lower_left_x % size_x, lower_left_y - lower_left_y % size_y),
+                     bbox.get_upper_right()], crs=bbox.crs)
+
+    def _make_split(self):
+        """ Split each UTM grid into equally sized bboxes in correct UTM zone
+        """
+        size_x, size_y = self.bbox_size
+        self.bbox_list = []
+        self.info_list = []
+
+        index = 0
+
+        for utm_cell in self.utm_grid:
+            utm_cell_geom, utm_cell_prop = utm_cell
+            if utm_cell_prop['zone'] == 0:
+                continue
+            utm_crs = self._get_utm_from_props(utm_cell_prop)
+
+            intersection = utm_cell_geom.intersection(self.shape_geometry.geometry)
+
+            if isinstance(intersection, (Polygon, MultiPolygon)):
+                bbox_utm = Geometry(utm_cell_geom, crs=CRS.WGS84).transform(utm_crs).bbox
+                bbox_utm = self._align_bbox_to_size(bbox_utm)
+                bbox_partition = bbox_utm.get_partition(size_x=size_x, size_y=size_y)
```

Comment:
In many cases intersection geometry would only cover a small part of entire UTM grid tile or zone. So it would be more efficient if only intersection bbox would be tiled and not entire UTM bbox.

I suggest something like this:
```
bbox_utm = Geometry(utm_cell_geom, crs=CRS.WGS84).transform(utm_crs).bbox
intersection = Geometry(intersection, CRS.WGS84).transform(utm_crs)

base_point = 0, 0 if isinstance(self, UtmZoneSplitter) else 0, bbox_utm.lower_left[1]
aligned_bbox = self._align_bbox(intersection.bbox, base_point)

bbox_partition = aligned_bbox.get_partition(size_x=size_x, size_y=size_y)
```
where I would change alignment method into
```Python
def _align_bbox(self, bbox, base_point):
        size_x, size_y = self.bbox_size
        lower_left_x, lower_left_y = bbox.lower_left
        base_x, base_y = base_point
        return BBox([(math.floor((lower_left_x - base_x) / size_x) * size_x,
                      math.floor((lower_left_y - base_y) / size_y) * size_y),
                     bbox.upper_right], crs=bbox.crs)
```
Btw, I will check if that `0, bbox_utm.lower_left[1]` works as I imagined...

Reply:
Update: `0, bbox_utm.lower_left[1]` doesn't work as I hoped.

Reply:
Looks like there is no good definition of lower end of an UTM grid. Then I would leave the method for alignemnt unchanged.
Still aligning and partitioning should be done on bbox of intersection, not `bbox_utm`.

Reply:
yes, that makes sense. Since any bbox gets resampled to multiples of `size_x` and `size_y`, the partition can be applied directly to the BBox of the intersection

```python
bbox_partition = self._align_bbox_to_size(intersection.bbox).get_partition(size_x=size_x, size_y=size_y)
```

with no need to change `_align_bbox` and deleting the previous lines

```python
bbox_utm = Geometry(utm_cell_geom, crs=CRS.WGS84).transform(utm_crs).bbox
bbox_utm = self._align_bbox_to_size(bbox_utm)
bbox_partition = bbox_utm.get_partition(size_x=size_x, size_y=size_y)
```